src/miniscript/musig_key.rs

```diff
@@ -0,0 +1,102 @@
+//! Support for multi-signature keys
+use std;
+use std::str::FromStr;
+use std::{error, fmt};
+
+use crate::expression::Tree;
+use crate::MiniscriptKey;
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
+/// Enum for representing keys in miniscript
+pub enum KeyExpr<Pk: MiniscriptKey> {
+    /// Single-key (e.g pk(a), here 'a' is a single key)
+    SingleKey(Pk),
+
+    /// Collection of keys in used for multi-signature
+    MuSig(Vec<KeyExpr<Pk>>),
+}
+
+#[derive(Debug, PartialEq)]
+/// Errors related to KeyExpr
+pub enum KeyExprError {
+    /// Single key is empty string
+    SingleKeyEmptyError,
+
+    /// Parsing error for single key
+    SingleKeyParseError,
+}
+
+impl fmt::Display for KeyExprError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            KeyExprError::SingleKeyEmptyError => f.write_str("str for single key was empty"),
+            KeyExprError::SingleKeyParseError => f.write_str("not able to parse the single key"),
+        }
+    }
+}
+
+impl error::Error for KeyExprError {}
+
+impl<Pk: MiniscriptKey + FromStr> FromStr for KeyExpr<Pk> {
+    type Err = KeyExprError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (key_tree, _) = Tree::from_slice(s).unwrap();
+        fn expr_from_tree<Pk: MiniscriptKey + FromStr>(
```

Comment:
You should 

impl<Pk> FromTree for KeyExpr<Pk> instead of this local function

Reply:
Implemented FromTreef for KeyExpr.